### PR TITLE
[Relay][Training] Add gradient for Crossentropy

### PR DESCRIPTION
**Pull request by MarisaKirisame**
@vinx13 @junrushao1994 @SWu can you guys help review?

### CODE REVIEW

File: python/tvm/relay/testing/__init__.py [hunk 1]
```diff
@@ -56,11 +56,11 @@ def run_infer_type(expr):
     return run_opt_pass(expr, transform.InferType())
 
 
-def _np_randn_from_type(t, scale=1):
-    return (scale * np.random.randn(*(int(d) for d in t.shape))).astype(t.dtype)
+def _np_randn_from_type(t, scale=1, mean=0):
+    return (mean + (scale * np.random.randn(*(int(d) for d in t.shape)))).astype(t.dtype)
 
 
-def check_grad(func, inputs=None, eps=1e-6, atol=1e-5, rtol=1e-3):
```
Comment (vinx13) [inline python/tvm/relay/testing/__init__.py]: document the new argument `scale`

File: python/tvm/relay/op/nn/_nn.py [hunk 2]
```diff
@@ -717,3 +717,16 @@ def schedule_bitserial_dense(attrs, outputs, target):
 
 
 reg.register_pattern("nn.bitserial_dense", reg.OpPattern.OUT_ELEMWISE_FUSABLE)
+
+
+
```
Comment (vinx13) [inline python/tvm/relay/op/nn/_nn.py]: nit: remove extra blank line (only two are needed)

**Comment by MarisaKirisame** [comment]
@vinx13 I had address your comment. can you review again?

File: python/tvm/relay/op/nn/nn.py [hunk 3]
```diff
@@ -1621,3 +1621,7 @@ def bitserial_dense(data,
     """
     return _make.bitserial_dense(data, weight, units, data_bits, weight_bits,
                                  pack_dtype, out_dtype, unipolar)
+
+
+def cross_entropy(predictions, targets):
```
Comment (SWu) [inline python/tvm/relay/op/nn/nn.py]: This should have a docstring. You should also mention that this is cross-entropy without softmax, as many frameworks equate cross-entropy to cross-entropy from logits

**Comment by MarisaKirisame** [comment]
@vinx13 @tqchen what should it's schedule be?

**Comment by vinx13** [comment]
@MarisaKirisame The schedule should be injective, can you check if the CUDA schedule are properly called?

**Comment by MarisaKirisame** [comment]
@vinx13 how can I do that? I am not really familiar with tvm low level internal.

**Comment by vinx13** [comment]
@MarisaKirisame I will take a look

**Comment by vinx13** [comment]
@MarisaKirisame I guess this is caused by the use of reference. It makes fusion and scheduling difficult. But I didn't reproduce the error on master, can you try rebasing?

File: python/tvm/relay/op/nn/_nn.py [hunk 4]
```diff
@@ -745,3 +745,15 @@ def schedule_bitserial_dense(attrs, outputs, target):
 
 
 reg.register_pattern("nn.bitserial_dense", reg.OpPattern.OUT_ELEMWISE_FUSABLE)
+
+
+reg.register_schedule("nn.cross_entropy", schedule_injective)
```
Comment (vinx13) [inline python/tvm/relay/op/nn/_nn.py]: the schedule actually should be schedule_reduce (in relay.op._reduce)

File: tests/python/relay/test_op_grad_level10.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from tvm import relay
+from tvm.relay.testing import check_grad
+
+
+def test_crossentropy_grad():
```
Comment (vinx13) [inline tests/python/relay/test_op_grad_level10.py]: nit: test_cross_entropy_grad

**Comment by MarisaKirisame** [comment]
@vinx13 @tqchen it is still the same error. as I dont has a cuda I cannot reproduce.

**Comment by vinx13** [comment]
@MarisaKirisame see my comment, the schedule should be reduce

**Comment by vinx13** [comment]
ping @MarisaKirisame

**Comment by MarisaKirisame** [comment]
@vinx13 sorry, I was pushing training on a private branch. I had addressed the issues.

**Comment by MarisaKirisame** [comment]
@vinx13 @tqchen can anyone take a look? I didnt change resize in any way. this is blocking training.

**Comment by vinx13** [comment]
@MarisaKirisame might be a flaky case, can you restart the ci?

File: python/tvm/relay/op/nn/nn.py (hunk 3 above)
Comment (vinx13) [reply python/tvm/relay/op/nn/nn.py]: @MarisaKirisame can you react on @SWu 's comment (also put it in REGISTER_RELAY_OP section)

**Comment by MarisaKirisame** [comment]
@vinx13 I will restart it right now. Just FYI I also got the same error last time.

**Comment by vinx13** [comment]
@MarisaKirisame you can try increasing rtol of the failing test

**Comment by MarisaKirisame** [comment]
@vinx13 it now work.

File: python/tvm/relay/op/nn/nn.py (hunk 3 above)
Comment (vinx13) [reply python/tvm/relay/op/nn/nn.py]: ping @MarisaKirisame

**Comment by vinx13** [comment]
@MarisaKirisame https://github.com/dmlc/tvm/pull/3925/files#r330840227

**Comment by MarisaKirisame** [comment]
@vinx13 I had acted on the comment.
